test/src/ui/widget.spec.ts

```diff
@@ -0,0 +1,1792 @@
+        expect(msg.width).to.be(-1);
+      });
+
+      it('should have a `height` of `-1`', () => {
+        let msg = ResizeMessage.UnknownSize;
+        expect(msg.height).to.be(-1);
+      });
+
+    });
+
+    describe('#constructor()', () => {
+
+      it('should accept a width and height', () => {
+        let msg = new ResizeMessage(100, 100);
+        expect(msg instanceof ResizeMessage).to.be(true);
+      });
+
+    });
+
+    describe('#width', () => {
+
+      it('should be the width passed to the constructor', () => {
+        let msg = new ResizeMessage(100, 200);
+        expect(msg.width).to.be(100);
+      });
+
+      it('should be a read-only property', () => {
+        let msg = new ResizeMessage(100, 200);
+        expect(() => { msg.width = 200; }).to.throwError();
+      });
+
+    });
+
+    describe('#height', () => {
+
+      it('should be the height passed to the constructor', () => {
+        let msg = new ResizeMessage(100, 200);
+        expect(msg.height).to.be(200);
+      });
+
+      it('should be a read-only property', () => {
+        let msg = new ResizeMessage(100, 200);
+        expect(() => { msg.height = 200; }).to.throwError();
+      });
+
+    });
+
+  });
+
+  describe('Layout', () => {
+
+    describe('#iter()', () => {
+
+      it('should create an iterator over the widgets in the layout', () => {
+        let layout = new LogLayout;
+        let iter = layout.iter();
+        let child = iter.next();
+        expect(child instanceof LogWidget).to.be(true);
+      });
+
+    });
+
+    describe('#onLayoutChanged()', () => {
+
+      it('should be invoked when the layout is installed on its parent widget', () => {
+        let widget = new Widget();
+        let layout = new LogLayout();
+        widget.layout = layout;
+        expect(layout.methods.indexOf('onLayoutChanged')).to.not.be(-1);
+      });
+
+    });
+
+    describe('#onChildRemoved()', () => {
+
+      it("should be invoked when a child widget's `parent` property is set to `null`", () => {
+        let parent = new Widget();
+        let widget = new Widget();
+        let layout = new LogLayout();
+        parent.layout = layout;
+        widget.parent = parent;
+        widget.parent = null;
+        expect(layout.methods.indexOf('onChildRemoved')).to.not.be(-1);
+      });
+
+    });
+
+    describe('#dispose()', () => {
+
+      it('should dispose of the resource held by the widget', () => {
+        let widget = new Widget();
+        let layout = new LogLayout();
+        widget.layout = layout;
+        layout.dispose();
+        expect(layout.parent).to.be(null);
+      });
+
+      it('should be called automatically when the parent is disposed', () => {
+        let widget = new Widget();
+        let layout = new LogLayout();
+        widget.layout = layout;
+        widget.dispose();
+        expect(layout.parent).to.be(null);
+        expect(layout.isDisposed).to.be(true);
+      });
```

Comment:
it should also dispose of all child widgets, which means your custom layout needs to reimplement `dispose` to do that.